### PR TITLE
fmp4-validate enhancements - add HDR and MVHEVC validation

### PR DESCRIPTION
**Pull request by elv-serban**
- enhance mp4e to parse color-specific mp4 and hevc elementrary stream
- update mod mp4ff to pick up clli and mdvc support
- enhance fmp4-validate with HDR output and detailed info
- add -json mode

Example output:

```
go run ./cmd/fmp4-validate -hdr -info   ../avpipe/TST/bypass-s.mp4
parse warning: decode rors pos 2336401: read box body length 3675 does not match expected length 2065851754
hdr:
  track: id=1 codec=hvc1
  [OK] hvcC/SPS: hvcC profile=Main10(2) bitDepth=10/10 nalLength=4; SPS profile=Main10(2) bitDepth=10/10 VUI=BT.2020(9)/PQ/ST2084(16)/BT.2020 non-constant(9) range=limited
  [OK] colr: type=nclx primaries=BT.2020(9) transfer=PQ/ST2084(16) matrix=BT.2020 non-constant(9) range=limited
  [FAIL] clli: missing clli box
  [FAIL] mdcv: missing mdcv box
  [OK] HEVC SEI: MDCV G(8500,39850)B(6550,2300)R(35400,14600)WP(15635,16450)L(10000000,10) from hvcC,sample; CLLI maxCLL=0 maxFALL=0 from hvcC,sample; samplesScanned=48 nalusScanned=57
info:
  codec: hvc1
  level: 6.0(180)
  profile: Main10(2)
  bit_depth: 10/10
  color_primaries: BT.2020(9)
  transfer_characteristics: PQ/ST2084(16)
  matrix_coefficients: BT.2020 non-constant(9)
  color_range: limited
  mastering_display: G(8500,39850)B(6550,2300)R(35400,14600)WP(15635,16450)L(10000000,10)
  max_cll_fall: 0,0
  max_luma: 1023
  min_luma: 0
```
```json
go run ./cmd/fmp4-validate -hdr -info -json  ../avpipe/TST/bypass-s.mp4  | jq .
{
  "parse_warning": "decode rors pos 2336401: read box body length 3675 does not match expected length 2065851754",
  "hdr": {
    "track": {
      "id": 1,
      "codec": "hvc1"
    },
    "checks": [
      {
        "name": "hvcC/SPS",
        "ok": true,
        "status": "OK",
        "detail": "hvcC profile=Main10(2) bitDepth=10/10 nalLength=4; SPS profile=Main10(2) bitDepth=10/10 VUI=BT.2020(9)/PQ/ST2084(16)/BT.2020 non-constant(9) range=limited"
      },
      {
        "name": "colr",
        "ok": true,
        "status": "OK",
        "detail": "type=nclx primaries=BT.2020(9) transfer=PQ/ST2084(16) matrix=BT.2020 non-constant(9) range=limited"
      },
      {
        "name": "clli",
        "ok": false,
        "status": "FAIL",
        "detail": "missing clli box"
      },
      {
        "name": "mdcv",
        "ok": false,
        "status": "FAIL",
        "detail": "missing mdcv box"
      },
      {
        "name": "HEVC SEI",
        "ok": true,
        "status": "OK",
        "detail": "MDCV G(8500,39850)B(6550,2300)R(35400,14600)WP(15635,16450)L(10000000,10) from hvcC,sample; CLLI maxCLL=0 maxFALL=0 from hvcC,sample; samplesScanned=48 nalusScanned=57"
      }
    ]
  },
  "info": {
    "codec": "hvc1",
    "level": "6.0(180)",
    "profile": "Main10(2)",
    "bit_depth": "10/10",
    "color_primaries": "BT.2020(9)",
    "transfer_characteristics": "PQ/ST2084(16)",
    "matrix_coefficients": "BT.2020 non-constant(9)",
    "color_range": "limited",
    "mastering_display": "G(8500,39850)B(6550,2300)R(35400,14600)WP(15635,16450)L(10000000,10)",
    "max_cll_fall": "0,0",
    "max_luma": "1023",
    "min_luma": "0"
  }
}

```

### CODE REVIEW

**Comment by elv-peter** [comment]
# Code Review: PR #166 — Add HDR and color parameter validation

**Author:** Serban Simu  
**State:** Draft  
**Additions/Deletions:** +1426 / -9  
**Reviewers requested:** elv-gerald, elv-peter, elv-preethi, eponymous301, lukseven

---

## Overview

Three related additions:

1. **`mp4e/hdr.go`** (~1169 lines): Pure-Go HDR validation for HEVC MP4/fMP4 files. Parses and cross-checks `hvcC` (including SPS/VUI), `colr`, `clli`, `mdcv` boxes, and HEVC SEI messages (MDCV, CLLI, x265 encoding settings). Entry point: `ValidateHDR(file *mp4.File) (*HDRInfo, error)`.

2. **`mp4e/mp4.go`**: Extends `ValidateFmp4` with HEVC-aware random-access frame checking (replaces the AVC-only `ensureIDRFrame` with a codec-dispatching `ensureRandomAccessFrame`). Integrates `HDRInfo` into `Mp4Info`.

3. **`cmd/fmp4-validate/main.go`**: Adds `-hdr`, `-info`, `-json` flags to the CLI tool, wiring up `ValidateHDR` with text and JSON output modes. Updates `go.mod` to point `github.com/Eyevinn/mp4ff` at the Eluvio fork (`eluv-io/mp4ff v0.47.1-0.20260505172703`).

---

## Code Quality

### `mp4e/hdr.go`

**Structure is clean and well-factored.** The separation between raw data (`HDRInfo`, `HDRHvcCInfo`, etc.) and display/report types (`HDRFieldInfo`, `HDRReport`) is appropriate. `addCheck` and `addSEIError` are good helpers that keep the check-accumulation logic DRY.

**`validateHvcC`** correctly checks both `hvcC` config record and SPS independently, flagging cases where VUI/colour_description is absent. The cross-check logic (profile, bit depth, primaries, transfer, matrix, range) is correct for HDR10.

**`validateColr`** uses `findColr` (linear scan over `vse.Children`) rather than `vse.Colr` directly. This is intentional — the mp4ff upstream `VisualSampleEntryBox` doesn't expose `Colr` as a named field (it varies by codec), so the child scan is necessary. Worth a comment explaining why `vse.Colr` isn't used.

**SEI scanning** is bounded by `hdrSEISampleScanLimitSeconds = 120` (2 minutes of wall time). This prevents hanging on very long files but means SEI content after the scan limit is invisible. The limit is logged in the check output string (`sampleScanLimit=120s`), which is good.

**`appendLimited`** caps SEI parse errors at 5 to avoid blowing up the output. Good defensive choice.

**`x265EncodingSetting`** parses unregistered SEI from x265 to extract min/max luma. This is fragile (depends on x265's specific string format), but it's the only practical way to get these values from the SEI stream, and it's clearly scoped to x265-identified SEIs.

**`scanProgressiveSamples`** reads `file.Mdat.Data` directly (whole mdat in memory). For very large progressive MP4s this could be a memory issue. Since this tool is for diagnostic use and the scan is bounded by time, this is acceptable — but worth a note in the function comment.

**`hdrSEISampleScanLimitTicks`**: returns `(0, false)` if `Timescale == 0`, disabling the scan limit. The `withinHDRSampleScanLimit` function treats `scanLimit == 0` as "no limit", which means a file with a zero timescale scans all samples. Consider whether that's intended or if it should be an error.

### Naming concern

`HDRHvcCInfo` contains both hvcC config record fields and SPS/VUI fields, with `SPS`-prefixed names for the latter. This works but the struct is large (22 fields). Consider splitting into `HDRHvcCConfigInfo` and `HDRSPSInfo` at some point if this grows further.

### `mp4e/mp4.go`

**`ensureRandomAccessFrame` → `ensureAVCIDRFrame` + HEVC RAP check** is a clean improvement. The previous code called `avc.GetNalusFromSample` on HEVC data and relied on the "Not video?" error path to silently pass — this was always wrong and the new codec-aware dispatch fixes it properly.

**`detectVideoCodec`** scans `moov.Traks` for the first video track with a recognized sample entry. The operator precedence in:
```go
if stsd.AvcX != nil || stsd.Encv != nil && stsd.Encv.AvcC != nil {
```
is correct (Go evaluates `&&` before `||`) but is hard to read. Consider adding parentheses:
```go
if stsd.AvcX != nil || (stsd.Encv != nil && stsd.Encv.AvcC != nil) {
```
Same for the HEVC case on the next line.

**`else if` → `if` in the segment/fragment loop**: Two `else if` chains are changed to plain `if`. The SeqEnd and DtsEnd/PtsEnd assignments are now always evaluated rather than only in the else-if branch. This is a subtle correctness fix — previously a single-fragment segment would fail to set SeqEnd (the `else if` could never fire when `fragIdx == 0 && fragIdx == len-1`). Good catch. Worth a comment.

### `cmd/fmp4-validate/main.go`

The flag interaction logic at the top of `run()`:
```go
if opts.json && !opts.hdr && !opts.info {
    opts.info = true
}
if opts.info {
    opts.hdr = true
}
if opts.json {
    opts.hdr = true
}
```
is correct but somewhat tangled. The semantics are: `json` implies `info` implies `hdr`. Could simplify to:
```go
if opts.info || opts.json {
    opts.hdr = true
}
if opts.json {
    opts.info = true
}
```
Current behavior and simplified version are equivalent, but the simplified form is easier to reason about.

**`mp4.DecodeFile` is called in the HDR path** while `ValidateFmp4` also calls `mp4.DecodeFile` in the non-HDR path. If `-hdr` is specified, the file is decoded once. If neither `-hdr` nor `-info` is set, `ValidateFmp4` decodes it. There's no double-decode. Good.

**Error handling**: `if file == nil { return fmt.Errorf(...) }` — this is redundant with `ValidateHDR` already checking for nil file. Not harmful but slightly duplicates the guard.

### `go.mod`

The replace directive:
```
github.com/Eyevinn/mp4ff => github.com/eluv-io/mp4ff v0.47.1-0.20260505172703-d379abafa307
```
replaces the upstream `v0.51.0` with the Eluvio fork at a pseudo-version (`v0.47.1-0.20260505172703`). The base version `v0.47.1` is lower than the upstream `v0.51.0`, which means the fork diverged at 0.47.1 and carries forward independently. **This is expected given the eluvio/mp4ff fork**, but it means upstream improvements between 0.47.1 and 0.51.0 are not present. Document this tradeoff somewhere (e.g., the CLAUDE.md note about the mp4ff fork).

---

## Potential Issues

1. **`validateColr` checks HDR10 values unconditionally**: The function checks `colorPrimaries == 9 && transfer == 16 && matrix == 9 && !fullRange`. This fails for any non-HDR10 HEVC file (e.g., SDR BT.709). The function is named `validateColr` and is called from `ValidateHDR`, so the caller context is HDR-only — but the check result will confusingly show `[FAIL]` for the colr box on any SDR HEVC file passed to `-hdr`. Consider noting in the String() output or the function name that this validates against HDR10 specifically.

2. **`findHEVCVideoSampleEntry` returns only the first HEVC track**: Multi-track files (e.g., a file with both AVC and HEVC tracks) will only validate the first HEVC video track. `ValidateHDR` returns a single `*HDRInfo`. This is appropriate for the current use case but undocumented.

3. **`scanFragmentedSamples` requires `trex`**: If `trak.Tkhd == nil`, `trex` will be nil and the function returns an error. However, the check `if trak.Tkhd != nil && trex == nil` treats a nil `Tkhd` as not requiring `trex`, and then passes a nil `trex` to `frag.GetFullSamples(trex)`. Verify `GetFullSamples` handles nil trex correctly.

4. **`probeVideoColor` in `xc/mez_e2e_test.go`** (also in PR 167): ABR init segments passed to `probeVideoColor` are MP4 init segments, not playable files. `avpipe.Probe` should handle them, but confirm this works for fMP4 init-only files.

---

## Test Coverage

- `TestValidateHDRHEVCInit` covers the main validation path with a real init segment (`testdata/hevc-init.m4s`). Checks all 5 HDR check types, `FieldInfo()`, `InfoString()`, `Report()`, and JSON marshaling. Good coverage.
- `TestHDRFieldInfoUsesX265EncodingSettingsForLuma` unit-tests the x265 settings parser and luma extraction. Good.
- `TestWithinHDRSampleScanLimit` tests edge cases including `decodeTime < firstDecodeTime`. Good.
- `TestEnsureRandomAccessFrameCodecAware` tests the AVC/HEVC/unknown dispatch. Covers the cross-codec case (HEVC sample with AVC codec → error).
- **Missing**: No test for `detectVideoCodec`. Easily unit-testable.
- **Missing**: No test for `ValidateHDR` on a non-HEVC file (should gracefully return five FAIL checks, not error).
- **Missing**: No test for the CLI flags in `fmp4-validate` (`run` function). The existing tests in `main_test.go` presumably cover the non-HDR path; a test for `-hdr` / `-json` output format would be valuable.

---

## Summary

Solid implementation. The HDR validation logic is thorough — checking boxes, SPS/VUI, and SEI independently and cross-referencing them. The HEVC random-access fix is a genuine correctness improvement for `ValidateFmp4`. The `else if` → `if` fix in the segment loop is a subtle correctness fix worth calling attention to in the PR description.

Priority items before merge:
1. Add parentheses in `detectVideoCodec` for clarity.
2. Clarify that `validateColr` validates HDR10 specifically (naming or comment).
3. Verify `GetFullSamples(nil trex)` behavior.
4. Consider a CLI test for `-hdr` / `-json` output.
5. Document the mp4ff fork version in CLAUDE.md or go.mod comments.

File: mp4e/hdr.go [hunk 1]
```diff
@@ -0,0 +1,903 @@
...
+			return samplesScanned, err
+		}
+		offset, err := getChunkOffset(stbl, chunkNr)
+		if err != nil {
+			return samplesScanned, err
+		}
+		for sNr := sampleNrAtChunkStart; sNr < sampleNr; sNr++ {
+			offset += uint64(stbl.Stsz.GetSampleSize(sNr))
+		}
+		size := stbl.Stsz.GetSampleSize(sampleNr)
```
[the remark this answers is not in the document]
Comment (elv-serban) [reply mp4e/hdr.go]: I removed the progressive part since we don't need it.  So this function is no longer used.

File: mp4e/hdr.go [hunk 2]
```diff
@@ -0,0 +1,903 @@
...
+}
+
+func (h *HDRInfo) scanMediaSamples(file *mp4.File, moov *mp4.MoovBox, trak *mp4.TrakBox, lengthSize int, sps *hevc.SPS) (int, error) {
+	if lengthSize < 1 || lengthSize > 4 {
+		return 0, fmt.Errorf("unsupported NALU length size %d", lengthSize)
+	}
+	if file.IsFragmented() {
+		return h.scanFragmentedSamples(file, moov, trak, lengthSize, sps)
+	}
+	return h.scanProgressiveSamples(file, trak, lengthSize, sps)
```
[the remark this answers is not in the document]
Comment (elv-serban) [reply mp4e/hdr.go]: I removed the progressive part since we don't need it

File: go.sum [hunk 3]
```diff
@@ -30,6 +28,8 @@ github.com/eluv-io/gls v1.0.0-elv1 h1:fV9z/aLSs5KH8wWMkamoS/Xxl8wmXF76WMx8/N5XEr
 github.com/eluv-io/gls v1.0.0-elv1/go.mod h1:4rPCrom1ZvL5sqFMGiNbCr/QfDq3oljqXEBdstdge0o=
 github.com/eluv-io/log-go v1.0.1 h1:j+/vNR7IwPxzgiwOWOKffWevEGeMhTXZVK7448FUuQg=
 github.com/eluv-io/log-go v1.0.1/go.mod h1:zXLeleJ9LQRHdEOvm7xh1dBQ/t5MJurj7vgfdUmL4jc=
+github.com/eluv-io/mp4ff v0.47.1-0.20260505172703-d379abafa307 h1:g7s5bqVyvO9+7hJynBa8Bsitr4CcPxNZByjK91LG2xE=
+github.com/eluv-io/mp4ff v0.47.1-0.20260505172703-d379abafa307/go.mod h1:hJNUUqOBryLAzUW9wpCJyw2HaI+TCd2rUPhafoS5lgg=
```
Comment (elv-peter) [inline go.sum]: Why not base off version v0.51?

**Comment by elv-serban** [comment]
I fixed "3. scanFragmentedSamples requires trex" - fail if trex is nil (we don't need to support that case)

File: cmd/fmp4-validate/main.go [hunk 4]
```diff
@@ -69,6 +113,9 @@ Usage of %s:
 	}
 
 	opts := options{}
+	fs.BoolVar(&opts.hdr, "hdr", false, "validate and print HDR10 HEVC metadata")
+	fs.BoolVar(&opts.info, "info", false, "print detailed HDR fields; implies -hdr")
+	fs.BoolVar(&opts.json, "json", false, "print HDR output as JSON; implies -info")
```
Comment (elv-peter) [inline cmd/fmp4-validate/main.go]: This is kind of weird (to set `-hdr` if the other flags are set).

Also, I would try to automatically detect HDR - e.g. if any of the HDR-specific atoms exist - even if no flags are set

File: mp4e/hdr.go [hunk 5]
```diff
@@ -0,0 +1,1169 @@
...
+	Video HDRVideoInfo
+	HvcC  HDRHvcCInfo
+	Colr  HDRColrInfo
+	Clli  HDRClliInfo
+	Mdcv  HDRMdcvInfo
+	SEI   HDRSEIInfo
+
+	Checks []HDRCheck
+	Errors []string
+}
```
Comment (elv-peter) [inline mp4e/hdr.go:35]: Not great to mix general video/media info into the HDR structs.

Also mixing media info reporting with validation should be avoided, except for contextual info relevant to debugging validation errors

File: mp4e/hdr.go [hunk 6]
```diff
@@ -0,0 +1,1169 @@
...
+func (h *HDRInfo) validateMdcv(vse *mp4.VisualSampleEntryBox) {
+	if vse.Mdcv == nil {
+		h.addCheck("mdcv", false, "missing mdcv box")
+		return
+	}
+	h.Mdcv = mdcvFromBox(vse.Mdcv)
+	h.addCheck("mdcv", true, "%s", formatMdcv(h.Mdcv))
+}
+
+func (h *HDRInfo) validateSEI(file *mp4.File, moov *mp4.MoovBox, trak *mp4.TrakBox, vse *mp4.VisualSampleEntryBox, sps *hevc.SPS) {
```
Comment (elv-peter) [inline mp4e/hdr.go:652]: I believe the SEI signaling is optional - redundant with the `mdcv` and `clli` MP4 boxes - for VoD.

It's overly strict to fail if there are is no HDR SEI data. Can just report that it was not found (but not mark it as an error/warning)

**Comment by elv-peter** [comment]
Claude:

⏺ A few things the current validator doesn't check:                                                                                                                                    
                                                                                                                                                                                     
  Bitstream                                                                                                                                                                            
  - Chroma sample location — HDR10 should have chroma_sample_loc_type_top_field = 2 (center-left) in the VUI. Wrong chroma siting causes subtle color fringing, especially noticeable  
  on fine edges.                                                                                                                                                                       
  - Video format in VUI (video_format field) — should be unspecified (5) for HDR10 delivery.                                                                                         
                                                                                                                                                                                       
  MP4 boxes                                                                                                                                                                            
  - mdcv and clli values are present but zero — the validator checks the boxes exist but doesn't validate the values are non-zero or within reasonable ranges. MaxCLL=0, MaxFALL=0     
  passes the check even though those are placeholder values that many encoders write when the actual analysis wasn't done. (The SEI cross-check does surface this via the x265 min/max 
  luma extraction, but only as informational.)                                                                                                                                         
  - mdcv primary/white point values are in the right ballpark — P3 primaries passing as BT.2020 would currently go undetected if the colr box says BT.2020 but the mdcv has P3         
  coordinates.                                                                                                                                                                       
  - pasp (pixel aspect ratio) — should be 1:1 for HDR10. Non-square pixels with HDR10 is unusual and probably wrong.                                                                   
   
  Cross-checks                                                                                                                                                                         
  - colr vs VUI consistency — the validator checks each independently against HDR10 constants, but doesn't explicitly flag when colr and VUI disagree with each other (e.g. colr says
  BT.2020 but VUI says BT.709). Both could individually pass partial checks while being inconsistent.                                                                                  
  - Pixel format — 10-bit YUV 4:2:0 is required for HDR10. The hvcC check verifies bit depth but not chroma format (4:2:0 vs 4:2:2 vs 4:4:4).                                        
                                                                                                                                                                                       
  The most practically important missing check is probably the mdcv/clli value sanity check — zero values are a common encoding artifact that silently passes.

**Comment by elv-peter** [comment]
Might be worth thinking about being able to do all validation in a single run of the tool

**Review by elv-peter — APPROVED** [review]

File: go.sum (hunk 3 above)
Comment (elv-serban) [reply go.sum]: I am based on 0.51

I need to:
- see why the dep doesn't show 0.51
- merge 0.52 - it's complicated by the fact they implemented some of the boxes (VPS) I added and I need to make sure theirs work!

**Comment by elv-serban** [comment]
### CLI spec

Always:
- the standard fmp4-validation

Sepecific validation:

- `-hdr`          (may have more 'specific' flags)
- `-mvhevc`
- `-atmos`

Options:
- `-info`  `-json`  print extra info  (will depend on the other flags like -hdr, -mvhevc, -atmos)

./fmp4-validate   file.mp4
./fmp4-validate  -json  file.mp4   
./fmp4-validate  -hdr  -mvhevc file.mp4

Modifiers:
-frames   (per frame stuff)
-idr           (idr frame info)

Output:
-json - print the same thing as JSON

**Review by elv-serban — COMMENTED** [review]
Looks good
